src/System.Private.CoreLib/shared/System/Threading/SpinWait.cs

```diff
@@ -127,12 +136,45 @@ internal set
         /// </remarks>
         public void SpinOnce()
         {
-            SpinOnce(DefaultSleep1Threshold);
+            SpinOnceCore(DefaultSleep1Threshold);
+        }
+
+        /// <summary>
+        /// Performs a single spin.
+        /// </summary>
+        /// <param name="sleep1Threshold">
+        /// A minimum spin count after which <code>Thread.Sleep(1)</code> may be used. A value of <code>-1</code> may be used to
+        /// disable the use of <code>Thread.Sleep(1)</code>.
+        /// </param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="sleep1Threshold"/> is less than <code>-1</code>.
+        /// </exception>
+        /// <remarks>
+        /// This is typically called in a loop, and may change in behavior based on the number of times a
+        /// <see cref="SpinOnce"/> has been called thus far on this instance.
+        /// </remarks>
+        public void SpinOnce(int sleep1Threshold)
```

Comment:
I've been thinking about this more, and I wonder whether we want to change the name of the argument to divorce it more from the implementation detail.  This is really about whether we want to allow SpinWait to block for a measurable period of time, and it just so happens we implement that using Thread.Sleep(1) as the smallest unit of blocking we can muster.  But it's quite possible in the future we'd have some other mechanism, and we would want the option of switching to that.  Maybe this should just be blockingThreshold, and "A minimum spin count after which measurable periods of blocking may be used", or something like that?  Or even "sleepThreshold", tying it more to the general notion of a thread sleeping rather than to the actual Thread.Sleep API?

Reply:
Sleep(0)/Yield() may wait for much longer when some threads are waiting to run. The Sleep(1) is an intended long wait even when there are no other threads waiting to run. Not sure if that distinction can be made clear in the name/description without implying that a spin would not block for a long duration before this threshold. I'll think on it some more.

Reply:
Maybe we can just expose both thresholds, yieldThreshold and longWaitThreshold?

Reply:
Then again, on a single-proc system sleep/yield can't be avoided

Reply:
Sleep(1) is probably the longest wait SpinWait would do. If there are medium length waits available in the future (between Thread.SpinWait and Thread.Sleep), this parameter probably should not also prevent that from being used, if it matters it would be better to expose another threshold for that. So I'm fine with keeping this specific to Sleep(1) and keep the name as "sleep1Threshold". I'm also fine with a generic name like "longSleepThreshold", where "Sleep" could be replaced with "Wait", "Pause", "Yield", etc.

Reply:
> I'm also fine with a generic name like "longSleepThreshold", where "Sleep" could be replaced with "Wait", "Pause", "Yield", etc.

I think I'd prefer that over `sleep1Threshold`; that just seems overly constraining.  `longWaitThreshold` sounds good to me personally.

Reply:
Ok, I'll leave it as `sleep1Threshold`, as I described I think it's sufficient and appropriate

Reply:
Let me know if you feel strongly about this, otherwise I'd rather leave well enough alone

Reply:
Ok